rollout/canary.go

```diff
@@ -291,30 +303,39 @@ func (c *RolloutController) syncRolloutStatusCanary(olderRSs []*appsv1.ReplicaSe
 			logCtx.Info("New RS has successfully progressed")
 			newStatus.Canary.StableRS = newStatus.CurrentPodHash
 		}
-		newStatus = c.calculateRolloutConditions(r, newStatus, allRSs, newRS, currExp)
+		newStatus = c.calculateRolloutConditions(r, newStatus, allRSs, newRS, currExp, currArs)
 		return c.persistRolloutStatus(r, &newStatus, pointer.BoolPtr(false))
 	}
 
-	if completedCurrentCanaryStep(olderRSs, newRS, stableRS, currExp, r) {
+	stepArs, _ := analysisutil.FilterAnalysisRunsByRolloutType(currArs, v1alpha1.RolloutTypeStepLabel)
+	var currStepAr *v1alpha1.AnalysisRun
+	if len(stepArs) == 1 {
```

Comment:
This len == 1 check seems fragile? If it's expected to only have one value, then maybe the helper should be changed to just GetCurrentStepAnalysisRun() and it should return nil or the single run value.